src/main/java/com/datasonnet/header/Header.java

```diff
@@ -0,0 +1,123 @@
+    public static String DATASONNET_HEADER = "/** DataSonnet";
+    public static String DATASONNET_VERSION = "version";
+    public static String DATASONNET_OUTPUT = "output";
+    public static String DATASONNET_INPUT = "input";
+
+    public static String DATAFORMAT_PREFIX = "dataformat";
+    public static String DATAFORMAT_DEFAULT = "*";
+
+
+    private String version = "1.0";
+    private Map dataFormatParameters = Collections.emptyMap();
+
+    public static Header parseHeader(String dataSonnetDocument) throws HeaderParseException {
+        Header header = new Header();
+
+        if (dataSonnetDocument.trim().startsWith(DATASONNET_HEADER)) {
+            String headerCommentSection = dataSonnetDocument.substring(0, dataSonnetDocument.indexOf("*/"));
+            headerCommentSection = headerCommentSection.replace(DATASONNET_HEADER, "").replace("*/","");
+
+            JavaPropsMapper mapper = new JavaPropsMapper();
+            JavaPropsSchema schema = new JavaPropsSchema() {
+                class HeaderSplitter extends JPropPathSplitter
+                {
+                    protected final char _pathSeparatorChar = '.';
+
+                    public HeaderSplitter()
+                    {
+                        super(true);
+                    }
+
+                    @Override
+                    public JPropNode splitAndAdd(JPropNode parent,
+                                                 String key, String value)
+                    {
+                        JPropNode curr = parent;
+                        String[] segments = key.split("(?<!" + Pattern.quote("\\") + ")" + Pattern.quote("" + _pathSeparatorChar));
+                        for (String segment : segments) {
+                            curr = _addSegment(curr, segment.replaceAll("\\\\", ""));
+                        }
+                        return curr.setValue(value);
+                    }
+                }
+
+                @Override
+                public JPropPathSplitter pathSplitter() {
+                    return new HeaderSplitter();
+                };
+            };
+
+            try {
+                Properties props = new Properties();
+                props.load(new StringReader(headerCommentSection));
+                Map propsMap = mapper.readPropertiesAs(props, schema, Map.class);
+                if (propsMap.containsKey(DATASONNET_VERSION)) {
+                    header.setVersion((String)propsMap.get(DATASONNET_VERSION));
+                }
+                header.dataFormatParameters = propsMap;
+            } catch (Exception e) {
+                throw new HeaderParseException("Error parsing DataSonnet Header: ", e);
+            }
+        }
+
+        return header;
+    }
+
+    public String getVersion() {
+        return version;
+    }
+
+    public void setVersion(String version) {
+        this.version = version;
+    }
+
+    public Map<String, Map<String, Object>> getDefaultParameters(String mimeType) {
```

Comment:
This is the wrong signature here, at the mimeType level the Map is (String) parameter names to objects (which might be strings or other maps or whatnot).

This code is still pretty opaque to me. I can read it, but I have to step through it very carefully to be sure there aren't bugs. Something like the method layout I recommended last time would result in a larger number of much shorter methods that are easier to check (and also test, if we put some of them in a utility class and added unit tests for example).  (I'm leaving out most type parameters for simplicity writing it, but they should be added everywhere; using the unmodifiable bits of Collections might be good too).

```java

Map getOrEmpty(Map<K, Map<>> map, K key) {
    return map.getOrDefault(key, Collections.emptyMap());
}

Map getDefaultParameters(Map parent) {
    return getOrEmpty(parent, Header.DATAFORMAT_DEFAULT);
}

Map overlay(Map defaults, Map overlay) {
    Map output = new HashMap(defaults);
    output.putAll(overlay);
    return output;
}
Map getInputParameters(String name) {
    Map inputs = getOrEmpty(dataFormatParameters, Header.DATASONNET_INPUT);
    Map inputDefaults = getDefaultParameters(inputs);
    return overlay(inputDefaults, getOrEmpty(inputs, name));
}

Map getInputParameters(String name, String mimetype) {
    return overlay(
        getOrEmpty(getDefaultParameters(dataFormatParameters), mimeType),
        getOrEmpty(getInputParameters(name), mimeType)
    );
}

Map getOutputParameters(String mimeType) {
    return overlay(
        getOrEmpty(getDefaultParameters(dataFormatParameters), mimeType),
        getOrEmpty(dataFormatParameters, Header.DATASONNET_OUTPUT)
    );
}


```

Reply:
OK, how about this one, not exactly as you described it but close.